src/compose/service.ts

```diff
@@ -560,22 +583,54 @@ export class Service {
 			tty: container.Config.Tty || false,
 		};
 
-		svc.appId = checkInt(svc.config.labels['io.balena.app-id']) || null;
-		svc.serviceId = checkInt(svc.config.labels['io.balena.service-id']) || null;
+		const appId = checkInt(svc.config.labels['io.balena.app-id']);
+		if (appId == null) {
+			throw new InternalInconsistencyError(
+				`Found a service with no appId! ${svc}`,
+			);
+		}
+		svc.appId = appId;
 		svc.serviceName = svc.config.labels['io.balena.service-name'];
+		svc.serviceId = parseInt(svc.config.labels['io.balena.service-id'], 10);
+		if (Number.isNaN(svc.serviceId)) {
+			throw new InternalInconsistencyError(
+				'Attempt to build Service class from container with malformed labels',
+			);
+		}
 		const nameMatch = container.Name.match(/.*_(\d+)_(\d+)$/);
+		if (nameMatch == null) {
+			throw new InternalInconsistencyError(
+				'Attempt to build Service class from container with malformed name',
+			);
+		}
 
-		svc.imageId = nameMatch != null ? checkInt(nameMatch[1]) || null : null;
-		svc.releaseId = nameMatch != null ? checkInt(nameMatch[2]) || null : null;
+		svc.imageId = parseInt(nameMatch[1], 10);
+		svc.releaseId = parseInt(nameMatch[2], 10);
 		svc.containerId = container.Id;
 
 		return svc;
 	}
 
-	public toComposeObject(): ServiceConfig {
-		// This isn't techinically correct as we do some changes
-		// to the configuration which we cannot reverse. We also
-		// represent the ports as a class, which isn't ideal
+	public toComposeObject() /*: TargetState['local']['apps'][0]['services'][0]*/ {
+		// Here we try to reverse the fromComposeObject to the best of our ability, as
+		// this is used for the supervisor reporting it's own target state. Some of
+		// these values won't match in a 1-1 comparison, such as `devices`, as we lose
+		// some data about
+		// return {
+		// 	...this.config,
+		// 	imageId: this.imageId!,
+		// 	serviceName: this.serviceName!
+		// 	devices: this.config.devices.map(composeUtils.dockerDeviceToStr),
+		// 	healthcheck: {
+		// 		...this.config.healthcheck,
+		// 		interval: this.config.healthcheck.interval?.toString(),
+		// 		timeout: this.config.healthcheck.timeout?.toString(),
+		// 		startPeriod: this.config.healthcheck.startPeriod?.toString(),
+		// 	},
+		// 	stopGracePeriod: this.config.startGracePeriod?.toString(),
+		// 	memLimit: this.config.memLimit?.toString(),
+		// 	memReservation: this.config.memReservation?.toString(),
+		// };
```

Comment:
I think most of this comment could be moved into jsdoc for the function where it'd be available for/show up in intellisense